File: Locomotor.swift
// MaleCNS v1.0 nerve-cord circuit. Anatomy and synapse counts are measured;
// LIF parameters, rate transfer between specimens, sensory tuning and muscle
// activation are explicit modeling assumptions (see data/LOCOMOTOR_PROVENANCE.md).
import Foundation

struct LocomotorNeuronFile: Decodable {
    let id: String
    let type: String
    let role: String
    let side: String
    let leg: Int?
    let motorChannel: String?
    let sensoryKind: String?
}

struct LocomotorCircuitFile: Decodable {
    let neurons: [LocomotorNeuronFile]
    let edges: [[Float]]

    func validate() -> Bool {
        guard !neurons.isEmpty, !edges.isEmpty,
              Set(neurons.map { $0.id }).count == neurons.count else { return false }
        for nr in neurons {
            if let leg = nr.leg, !(0..<6).contains(leg) { return false }
        }
        for e in edges {
            guard e.count == 3, e.allSatisfy({ $0.isFinite }),
                  e[0] >= 0, e[1] >= 0, e[0] < Float(neurons.count),
                  e[1] < Float(neurons.count), e[0].rounded() == e[0],
                  e[1].rounded() == e[1] else { return false }
        }
        return (0..<6).allSatisfy { leg in
            ["tibia_flexor", "tibia_extensor", "trochanter_flexor", "trochanter_extensor"].allSatisfy { channel in
                neurons.contains { $0.leg == leg && $0.motorChannel == channel }
            }
        }
    }
}

struct LocomotorParameters {
    var synapticGain: Double = 2.4
    var baseline: Double = 0.022
    var adaptationKick: Double = 0.01
}

final class LocomotorSim {
    // The per-millisecond loops below run over ~1,000 neurons a thousand times a
    // second, so everything they touch is resolved to an integer index here at
    // init: a string compare, a dictionary hash, or a copy of a neuron struct
    // (six retained String fields) inside those loops costs more than the LIF
    // arithmetic it guards. Names still come from the data; only lookups change.
    static let motorChannelNames = [
        "coxa_promotor", "coxa_anterior_rotator", "coxa_remotor", "coxa_posterior_rotator",
        "trochanter_flexor", "trochanter_extensor", "tibia_flexor", "tibia_extensor",
    ]
    private enum Role: UInt8 { case other = 0, motor, sensory, descending, ascending }
    private enum SensoryKind: UInt8 { case load = 0, hairPlate, excursion }

    let circuit: LocomotorCircuitFile
    let n: Int
    private var voltage: [Double]
    private var adaptation: [Double]
    private var refractory: [Int]
    private var rates: [Double]
    private var excitatory: [Double]
    private var inhibitory: [Double]
    private var nextExcitatory: [Double]
    private var nextInhibitory: [Double]
    private var rowStart: [Int]
    private var targets: [Int]
    private var weights: [Double]
    private var drive: [Double]
    private var sensoryDrive: [Double]
    private var roleCode: [UInt8] = []
    private var commandGroupIds: [[Int]] = []
    private var commandGroupIndex: [String: Int] = [:]
    private var motorChannelIds: [[Int]] = []   // leg * motorChannelNames.count + channel
    private var sensory: [Int] = []
    private var sensoryLeg: [Int] = []          // parallel to `sensory`
    private var sensoryKind: [UInt8] = []       // parallel to `sensory`
    private(set) var commands = Array(repeating: LegMotorCommand(), count: 6)
    private(set) var totalSpikes = 0
    private(set) var motorSpikes = 0
    private(set) var sensorySpikes = 0
    private(set) var simMs = 0
    var feedback: [LegFeedback] = []
    // Lesions are diagnostic interventions, used to verify actual causal paths.
    // The set stays the API; the hot loop reads the mirrored flags instead, so a
    // normal run costs one Bool check rather than a hash per neuron per ms.
    var silenced: Set<Int> = [] {
        didSet {
            anySilenced = !silenced.isEmpty
            for i in 0..<n { silencedFlag[i] = silenced.contains(i) }
        }
    }
    private var silencedFlag: [Bool] = []
    private var anySilenced = false
    var synapsesEnabled = true
    var feedbackEnabled = true
    let parameters: LocomotorParameters

    init(circuit: LocomotorCircuitFile, parameters: LocomotorParameters = LocomotorParameters()) {
        precondition(circuit.validate(), "invalid MaleCNS locomotor circuit")
        self.circuit = circuit
        self.parameters = parameters
        n = circuit.neurons.count
        voltage = .init(repeating: 0, count: n)
        adaptation = .init(repeating: 0, count: n)
        refractory = .init(repeating: 0, count: n)
        rates = .init(repeating: 0, count: n)
        excitatory = .init(repeating: 0, count: n)
        inhibitory = .init(repeating: 0, count: n)
        nextExcitatory = .init(repeating: 0, count: n)
        nextInhibitory = .init(repeating: 0, count: n)
        drive = .init(repeating: 0, count: n)
        sensoryDrive = .init(repeating: 0, count: n)
        silencedFlag = .init(repeating: false, count: n)
        roleCode = circuit.neurons.map {
            switch $0.role {
            case "motor": return Role.motor.rawValue
            case "sensory": return Role.sensory.rawValue
            case "descending": return Role.descending.rawValue
            case "ascending": return Role.ascending.rawValue
            default: return Role.other.rawValue
            }
        }
        motorChannelIds = Array(repeating: [], count: 6 * Self.motorChannelNames.count)
        var counts = Array(repeating: 0, count: n)
        var inputTotal = Array(repeating: Double(0), count: n)
        for e in circuit.edges {
            counts[Int(e[0])] += 1
            inputTotal[Int(e[1])] += abs(Double(e[2]))
        }
        rowStart = Array(repeating: 0, count: n + 1)
        for i in 0..<n { rowStart[i + 1] = rowStart[i] + counts[i] }
        targets = Array(repeating: 0, count: circuit.edges.count)
        weights = Array(repeating: 0, count: circuit.edges.count)
        var fill = rowStart
        for e in circuit.edges {
            let pre = Int(e[0]), post = Int(e[1]), slot = fill[pre]
            targets[slot] = post
            // Normalize retained input to avoid equating synapse count with a
            // measured conductance. Relative counts and transmitter signs survive.
            weights[slot] = parameters.synapticGain * Double(e[2]) / max(60, inputTotal[post])
            fill[pre] += 1
        }
        for (i, nr) in circuit.neurons.enumerated() {
            if nr.role == "descending" {
                let key = "\(nr.type):\(nr.side)"
                let group: Int
                if let existing = commandGroupIndex[key] { group = existing }
                else {
                    group = commandGroupIds.count
                    commandGroupIndex[key] = group
                    commandGroupIds.append([])
                }
                commandGroupIds[group].append(i)
            }
            if nr.role == "sensory", let leg = nr.leg {
                sensory.append(i)
                sensoryLeg.append(leg)
                // Same three-way split the step loop used to spell out with
                // string compares; the pooling rationale is unchanged.
                let kind: SensoryKind
                if nr.sensoryKind == "campaniform" || nr.sensoryKind == "contact" { kind = .load }
                else if nr.sensoryKind == "hair_plate" { kind = .hairPlate }
                else { kind = .excursion }
                sensoryKind.append(kind.rawValue)
            }
            if nr.role == "motor", let leg = nr.leg, let channel = nr.motorChannel,
               let slot = Self.motorChannelNames.firstIndex(of: channel) {
                motorChannelIds[leg * Self.motorChannelNames.count + slot].append(i)
            }
        }
    }

    // A modeled homologous population-rate interface between female FlyWire
    // and male CNS specimens. It adds current, never fabricated graph edges.
    func setDescending(_ type: String, side: String, rate: Float) {
        guard let group = commandGroupIndex["\(type):\(side)"] else { return }
        setDescending(group: group, rate: rate)
    }

    // Index form of the above, for the caller that resolves its groups once.
    func descendingGroup(_ type: String, side: String) -> Int? {
        commandGroupIndex["\(type):\(side)"]
    }

    func setDescending(group: Int, rate: Float) {
        let value = min(0.35, max(0, Double(rate)) * 0.004)
        for i in commandGroupIds[group] { drive[i] = value }
    }

    func meanRate(role: String, leg: Int? = nil) -> Float {
        let ids = indices(role: role, leg: leg)
        return Float(ids.reduce(0) { $0 + rates[$1] } / Double(max(1, ids.count)))
    }

    func indices(role: String, leg: Int? = nil) -> [Int] {
        guard let code = Self.roleCode(role) else { return [] }
        return (0..<n).filter {
            roleCode[$0] == code && (leg == nil || circuit.neurons[$0].leg == leg)
        }
    }

    private static func roleCode(_ role: String) -> UInt8? {
        switch role {
        case "motor": return Role.motor.rawValue
        case "sensory": return Role.sensory.rawValue
        case "descending": return Role.descending.rawValue
        case "ascending": return Role.ascending.rawValue
        case "premotor": return Role.other.rawValue
        default: return nil
        }
    }

    func step(_ ms: Int) {
        guard ms > 0 else { return }
        for _ in 0..<ms {
            simMs += 1
            // Leg-local sensory transduction, without global gait phase or
            // neuron-ID-derived tuning. Missing direction tuning is pooled:
            // proprioceptors encode joint excursion/speed; contact sensors load.
            for k in sensory.indices { sensoryDrive[sensory[k]] = 0 }
            if feedbackEnabled && feedback.count == 6 {
                for k in sensory.indices {
                    let f = feedback[sensoryLeg[k]]
                    let value: CGFloat
                    switch sensoryKind[k] {
                    case SensoryKind.load.rawValue:
                        value = f.contact ? min(1, f.load * 6) : 0
                    case SensoryKind.hairPlate.rawValue:
                        value = min(1, abs(f.hipAngle) / LegDynamics.hipLimit
                                    + abs(f.elevationVelocity) / 20)
                    default:
                        value = min(1, abs(f.kneeVelocity) / 20 + abs(f.hipVelocity) / 16
                                      + abs(f.kneeAngle - LegDynamics.restKnee) * 0.35)
                    }
                    sensoryDrive[sensory[k]] = Double(value) * 0.10
                }
            }
            for i in 0..<n {
                // Finite synaptic currents, with 5 ms excitatory and 10 ms
                // inhibitory decay. These timescales are model parameters,
                // not measured for the reconstructed specimen.
                excitatory[i] = excitatory[i] * 0.8187308 + nextExcitatory[i]
                inhibitory[i] = inhibitory[i] * 0.9048374 + nextInhibitory[i]
                nextExcitatory[i] = 0; nextInhibitory[i] = 0
            }
            for i in 0..<n {
                rates[i] *= 0.9048374 // 10 ms rate time constant for fast muscles
                adaptation[i] *= 0.9950125 // 200 ms spike-frequency adaptation
                if anySilenced && silencedFlag[i] {
                    voltage[i] = 0; rates[i] = 0; continue
                }
                if refractory[i] > 0 { refractory[i] -= 1; continue }
                // Deterministic subthreshold excitability, without autonomous
                // noise or gait oscillators. Input must recruit the real graph.
                voltage[i] = max(-1, voltage[i] * 0.9512294 + excitatory[i] + inhibitory[i]
                                     + parameters.baseline + drive[i] + sensoryDrive[i] - adaptation[i])
                if voltage[i] >= 1 {
                    voltage[i] = 0; refractory[i] = 2
                    adaptation[i] += parameters.adaptationKick
                    rates[i] += 95.16258
                    totalSpikes += 1
                    if roleCode[i] == Role.motor.rawValue { motorSpikes += 1 }
                    else if roleCode[i] == Role.sensory.rawValue { sensorySpikes += 1 }
                    if synapsesEnabled {
                        for e in rowStart[i]..<rowStart[i + 1] {
                            if weights[e] >= 0 { nextExcitatory[targets[e]] += weights[e] }
                            else { nextInhibitory[targets[e]] += weights[e] }
                        }
                    }
                }
            }
        }
        for leg in 0..<6 {
            let base = leg * Self.motorChannelNames.count
            func activity(_ slot: Int) -> CGFloat {
                let ids = motorChannelIds[base + slot]
                guard !ids.isEmpty else { return 0 }
                let rate = ids.reduce(0) { $0 + rates[$1] } / Double(ids.count)
                return CGFloat(rate / (rate + 50))
            }
            // Slots follow motorChannelNames: promotor, anterior rotator, remotor,
            // posterior rotator, trochanter flexor/extensor, tibia flexor/extensor.
            commands[leg] = LegMotorCommand(
                protract: max(activity(0), activity(1)),
                retract: max(activity(2), activity(3)),
                lift: activity(4), depress: activity(5),
                flex: activity(6), extend: activity(7))
        }
    }
}

File: windows/src/locomotor.js
// MaleCNS v1.0 nerve-cord circuit. Anatomy and synapse counts are measured;
// LIF parameters, rate transfer between specimens, sensory tuning and muscle
// activation are modeling assumptions. Mirrors Locomotor.swift.
import { LegDynamics, makeLegMotorCommand } from './legdynamics.js';

export function validateLocomotorCircuit(circuit) {
  if (!circuit || !Array.isArray(circuit.neurons) || !circuit.neurons.length
      || !Array.isArray(circuit.edges) || !circuit.edges.length) return false;
  const { neurons, edges } = circuit;
  if (new Set(neurons.map((neuron) => neuron.id)).size !== neurons.length) return false;
  for (const nr of neurons) {
    if (nr.leg != null && (!Number.isInteger(nr.leg) || nr.leg < 0 || nr.leg >= 6)) return false;
  }
  for (const e of edges) {
    if (!Array.isArray(e) || e.length !== 3 || !e.every(Number.isFinite)
        || !Number.isInteger(e[0]) || !Number.isInteger(e[1])
        || e[0] < 0 || e[1] < 0 || e[0] >= neurons.length || e[1] >= neurons.length) return false;
  }
  return Array.from({ length: 6 }, (_, leg) => leg).every((leg) =>
    ['tibia_flexor', 'tibia_extensor', 'trochanter_flexor', 'trochanter_extensor'].every((channel) =>
      neurons.some((nr) => nr.leg === leg && nr.motorChannel === channel)));
}

// Channel order is shared with Locomotor.swift's motorChannelNames; the step
// loop indexes it instead of hashing the names on every simulated millisecond.
const MOTOR_CHANNELS = [
  'coxa_promotor', 'coxa_anterior_rotator', 'coxa_remotor', 'coxa_posterior_rotator',
  'trochanter_flexor', 'trochanter_extensor', 'tibia_flexor', 'tibia_extensor',
];
const ROLE_OTHER = 0, ROLE_MOTOR = 1, ROLE_SENSORY = 2, ROLE_DESCENDING = 3, ROLE_ASCENDING = 4;
const SENSE_LOAD = 0, SENSE_HAIR_PLATE = 1, SENSE_EXCURSION = 2;
const ROLE_CODES = { motor: ROLE_MOTOR, sensory: ROLE_SENSORY, descending: ROLE_DESCENDING, ascending: ROLE_ASCENDING, premotor: ROLE_OTHER };

export class LocomotorSim {
  constructor(circuit, parameters = {}) {
    this.parameters = { synapticGain: 2.4, baseline: 0.022, adaptationKick: 0.01, ...parameters };
    if (!validateLocomotorCircuit(circuit)) throw new Error('Invalid MaleCNS locomotor circuit');
    this.circuit = circuit;
    this.n = circuit.neurons.length;
    const n = this.n;
    for (const field of ['voltage', 'adaptation', 'rates', 'excitatory', 'inhibitory', 'nextExcitatory', 'nextInhibitory', 'drive', 'sensoryDrive']) {
      this[field] = new Float64Array(n);
    }
    this.refractory = new Int32Array(n);
    // The loops below run over ~1,000 neurons a thousand times a second, so
    // everything they touch is reduced to an integer index here at construction:
    // a Map hash, a string compare or a neuron-object property load inside them
    // costs more than the LIF arithmetic it guards. Mirrors Locomotor.swift.
    this.commandGroupIndex = new Map();
    this.commandGroupIds = [];
    this.motorChannelIds = Array.from({ length: 6 * MOTOR_CHANNELS.length }, () => []);
    this.roleCode = new Uint8Array(n);
    this.sensory = [];
    this.sensoryLeg = [];
    this.sensoryKind = [];
    this.commands = Array.from({ length: 6 }, makeLegMotorCommand);
    this.totalSpikes = 0; this.motorSpikes = 0; this.sensorySpikes = 0; this.simMs = 0;
    this.feedback = [];
    this.silencedFlag = new Uint8Array(n);
    this.anySilenced = false;
    this._silenced = new Set();
    this.synapsesEnabled = true;
    this.feedbackEnabled = true;
    const counts = new Int32Array(n), inputTotal = new Float64Array(n);
    for (const [pre, post, weight] of circuit.edges) {
      counts[pre]++; inputTotal[post] += Math.abs(weight);
    }
    this.rowStart = new Int32Array(n + 1);
    for (let i = 0; i < n; i++) this.rowStart[i + 1] = this.rowStart[i] + counts[i];
    this.targets = new Int32Array(circuit.edges.length);
    this.weights = new Float64Array(circuit.edges.length);
    const fill = Int32Array.from(this.rowStart);
    for (const [pre, post, weight] of circuit.edges) {
      const slot = fill[pre]++;
      this.targets[slot] = post;
      // Keep relative counts and transmitter signs; counts are not conductance.
      this.weights[slot] = this.parameters.synapticGain * weight / Math.max(60, inputTotal[post]);
    }
    circuit.neurons.forEach((nr, i) => {
      this.roleCode[i] = ROLE_CODES[nr.role] ?? ROLE_OTHER;
      if (nr.role === 'descending') {
        const key = `${nr.type}:${nr.side}`;
        if (!this.commandGroupIndex.has(key)) {
          this.commandGroupIndex.set(key, this.commandGroupIds.length);
          this.commandGroupIds.push([]);
        }
        this.commandGroupIds[this.commandGroupIndex.get(key)].push(i);
      }
      if (nr.role === 'sensory' && nr.leg != null) {
        this.sensory.push(i);
        this.sensoryLeg.push(nr.leg);
        // Same three-way split the step loop used to spell out with string
        // compares; the pooling rationale is unchanged.
        this.sensoryKind.push(nr.sensoryKind === 'campaniform' || nr.sensoryKind === 'contact'
          ? SENSE_LOAD
          : nr.sensoryKind === 'hair_plate' ? SENSE_HAIR_PLATE : SENSE_EXCURSION);
      }
      if (nr.role === 'motor' && nr.leg != null && nr.motorChannel) {
        const slot = MOTOR_CHANNELS.indexOf(nr.motorChannel);
        if (slot >= 0) this.motorChannelIds[nr.leg * MOTOR_CHANNELS.length + slot].push(i);
      }
    });
  }

  // Lesions are diagnostic interventions, used to verify actual causal paths.
  // The Set stays the API; the hot loop reads the mirrored flags instead, so a
  // normal run costs one boolean check rather than a hash per neuron per ms.
  get silenced() { return this._silenced; }

  set silenced(value) {
    this._silenced = value instanceof Set ? value : new Set(value);
    this.anySilenced = this._silenced.size > 0;
    for (let i = 0; i < this.n; i++) this.silencedFlag[i] = this._silenced.has(i) ? 1 : 0;
  }

  setDescending(type, side, rate) {
    const group = this.descendingGroup(type, side);
    if (group >= 0) this.setDescendingGroup(group, rate);
  }

  // Index form of the above, for the caller that resolves its groups once.
  descendingGroup(type, side) {
    const group = this.commandGroupIndex.get(`${type}:${side}`);
    return group === undefined ? -1 : group;
  }

  setDescendingGroup(group, rate) {
    const value = Math.min(0.35, Math.max(0, rate) * 0.004);
    for (const i of this.commandGroupIds[group]) this.drive[i] = value;
  }

  indices(role, leg = null) {
    const code = ROLE_CODES[role];
    if (code === undefined) return [];
    const out = [];
    for (let i = 0; i < this.n; i++) {
      if (this.roleCode[i] === code && (leg === null || this.circuit.neurons[i].leg === leg)) out.push(i);
    }
    return out;
  }

  meanRate(role, leg = null) {
    const ids = this.indices(role, leg);
    return ids.reduce((sum, i) => sum + this.rates[i], 0) / Math.max(1, ids.length);
  }

  step(ms) {
    if (!(ms > 0)) return;
    for (let t = 0; t < ms; t++) {
      this.simMs++;
      for (const i of this.sensory) this.sensoryDrive[i] = 0;
      if (this.feedbackEnabled && this.feedback.length === 6) {
        for (let k = 0; k < this.sensory.length; k++) {
          const f = this.feedback[this.sensoryLeg[k]];
          const kind = this.sensoryKind[k];
          const value = kind === SENSE_LOAD
            ? (f.contact ? Math.min(1, f.load * 6) : 0)
            : kind === SENSE_HAIR_PLATE
              ? Math.min(1, Math.abs(f.hipAngle) / LegDynamics.hipLimit + Math.abs(f.elevationVelocity) / 20)
              : Math.min(1, Math.abs(f.kneeVelocity) / 20 + Math.abs(f.hipVelocity) / 16
              + Math.abs(f.kneeAngle - LegDynamics.restKnee) * 0.35);
          this.sensoryDrive[this.sensory[k]] = value * 0.10;
        }
      }
      for (let i = 0; i < this.n; i++) {
        this.excitatory[i] = this.excitatory[i] * 0.8187308 + this.nextExcitatory[i];
        this.inhibitory[i] = this.inhibitory[i] * 0.9048374 + this.nextInhibitory[i];
        this.nextExcitatory[i] = 0; this.nextInhibitory[i] = 0;
      }
      for (let i = 0; i < this.n; i++) {
        this.rates[i] *= 0.9048374;
        this.adaptation[i] *= 0.9950125;
        if (this.anySilenced && this.silencedFlag[i]) { this.voltage[i] = 0; this.rates[i] = 0; continue; }
        if (this.refractory[i] > 0) { this.refractory[i]--; continue; }
        this.voltage[i] = Math.max(-1, this.voltage[i] * 0.9512294 + this.excitatory[i] + this.inhibitory[i]
          + this.parameters.baseline + this.drive[i] + this.sensoryDrive[i] - this.adaptation[i]);
        if (this.voltage[i] >= 1) {
          this.voltage[i] = 0; this.refractory[i] = 2;
          this.adaptation[i] += this.parameters.adaptationKick;
          this.rates[i] += 95.16258;
          this.totalSpikes++;
          if (this.roleCode[i] === ROLE_MOTOR) this.motorSpikes++;
          else if (this.roleCode[i] === ROLE_SENSORY) this.sensorySpikes++;
          if (this.synapsesEnabled) {
            for (let e = this.rowStart[i]; e < this.rowStart[i + 1]; e++) {
              if (this.weights[e] >= 0) this.nextExcitatory[this.targets[e]] += this.weights[e];
              else this.nextInhibitory[this.targets[e]] += this.weights[e];
            }
          }
        }
      }
    }
    for (let leg = 0; leg < 6; leg++) {
      const base = leg * MOTOR_CHANNELS.length;
      const activity = (slot) => {
        const ids = this.motorChannelIds[base + slot];
        if (!ids.length) return 0;
        const rate = ids.reduce((sum, i) => sum + this.rates[i], 0) / ids.length;
        return rate / (rate + 50);
      };
      // Slots follow MOTOR_CHANNELS: promotor, anterior rotator, remotor,
      // posterior rotator, trochanter flexor/extensor, tibia flexor/extensor.
      this.commands[leg] = { protract: Math.max(activity(0), activity(1)),
        retract: Math.max(activity(2), activity(3)),
        lift: activity(4), depress: activity(5),
        flex: activity(6), extend: activity(7) };
    }
  }
}
